### PR TITLE
Feat/push image ghcr

### PR DESCRIPTION
**Pull request by ErnstWevers**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,7 +10,7 @@
     "build:accept": "yarn esbuild:generic accept/src/index.ts --outfile=accept/dist/index.js",
     "build:commit-status": "yarn esbuild:generic commit-status/src/index.ts --outfile=commit-status/dist/index.js",
     "build:db": "yarn esbuild:generic db/src/index.ts --outfile=db/dist/index.js",
-    "build:deploy-image": "yarn esbuild:generic deploy-image/src/index.ts --outfile=deploy-image/dist/index.js",
+    "build:push-image-ghcr": "yarn esbuild:generic push-image-ghcr/src/index.ts --outfile=push-image-ghcr/dist/index.js",
```
Comment (hongaar) [inline package.json:13]: nit: order alphabetically

File: README.md [hunk 2]
```diff
@@ -8,7 +8,7 @@ _Available actions:_
 - [`accept`](#accept)
 - [`commit-status`](#commit-status)
 - [`db`](#db)
-- [`deploy-image`](#deploy-image)
+- [`push-image-ghcr`](#push-image-ghcr)
```
Comment (hongaar) [inline README.md:11]: nit: order alphabetically

File: README.md [hunk 3]
```diff
@@ -102,35 +102,29 @@ repository migrations and runs them.
...
-| `docker-hub-user`     | ✅       |                    | Username for Docker Hub                                         |
-| `docker-hub-password` | ✅       |                    | Password for Docker Hub                                         |
+
 | `ghcr-user`           |          | `github.actor`     | Username for GitHub Container Registry                          |
 | `ghcr-password`       |          | `env.GITHUB_TOKEN` | Password for GitHub Container Registry                          |
 | `dockerfile`          |          | `"./Dockerfile"`   | Path to the Dockerfile                                          |
-| `gh-token`            |          | `env.GITHUB_TOKEN` | The github token to delete image tags (only for 'delete' event) |
-
-### `dry-run`
+| `gh-token`            |          | `env.GITHUB_TOKEN` |
```
Comment (hongaar) [inline README.md:126]:
```suggestion
| `gh-token` | | `env.GITHUB_TOKEN` | The github token to delete image tags (only for 'delete' event) |
```

File: push-image-ghcr/action.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-name: deploy-image
+name: push-image-ghcr
 description: Builds and deploys a component image
 inputs:
```
Comment (hongaar) [inline push-image-ghcr/action.yml:3]: please align with inputs in README

File: push-image-ghcr/src/index.ts [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import {
+  setComponentVersion,
+  setLabels,
+  setTags,
+} from './metadata'
+
+const METADATA_FILENAME = 'metadata.json'
+
+async function run() {
+  const { component: defaultComponent } = getRepository()
```
Comment (hongaar) [inline push-image-ghcr/src/index.ts:17]: please align with inputs in README

File: push-image-ghcr/src/index.ts [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  const component = getInput('component') || defaultComponent
+  const ghcrUser = getInput('ghcr-user') || context.actor
+  const ghcrPassword = getInput('ghcr-password') || getToken()
+  const dockerfile = getInput('dockerfile') || './Dockerfile'
+  const private_key = process.env.PRIVATE_KEY
```
Comment (hongaar) [inline push-image-ghcr/src/index.ts:22]: unused

File: push-image-ghcr/src/metadata.ts [hunk 7]
```diff
@@ -0,0 +1,35 @@
+import { getExecOutput } from '@actions/exec'
+import { parse as semverParse } from 'semver'
+import { getRef, getSha, getTag } from '../../lib/github'
+
+export async function setTags() {
```
Comment (hongaar) [inline push-image-ghcr/src/metadata.ts:5]: get verb matches logic better

```suggestion
export async function getTags() {
```

File: push-image-ghcr/src/metadata.ts [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+  const ref = getRef()
+  const sha = await getSha()
+  return [ref, sha]
+}
+
+export function setLabels({
```
Comment (hongaar) [inline push-image-ghcr/src/metadata.ts:11]: idem: fn name

File: push-image-ghcr/src/metadata.ts [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    'org.opencontainers.image.description': component,
+    'org.opencontainers.image.url': 'https://exivity.com',
+    'org.opencontainers.image.documentation': 'https://docs.exivity.com',
+    'org.opencontainers.image.source': `https://github.com/${process.env['GITHUB_REPOSITORY']}`,
+    'org.opencontainers.image.created': new Date().toISOString(),
+    'org.opencontainers.image.revision': getSha(),
+  }
+}
+
+export function setComponentVersion() {
```
Comment (hongaar) [inline push-image-ghcr/src/metadata.ts:30]: idem: fn name

File: deploy-image/src/index.ts [hunk 10]
```diff
@@ -1,179 +0,0 @@
...
-          package_name: component,
-          package_version_id: version.id,
-        })
-      } else {
-        warning('Could not find matching package version to delete')
-      }
-    }
-
-    return
-  }
```
Comment (hongaar) [inline deploy-image/src/index.ts:99]: We'd need to extract this logic into the new action and document this. I can take on this part if you want

File: deploy-image/src/index.ts [hunk 11]
```diff
@@ -1,179 +0,0 @@
...
-    `Writing metadata to ${METADATA_FILENAME}:\n${JSON.stringify(
-      metadata,
-      undefined,
-      2
-    )}`
-  )
-  await fs.writeFile(
-    './' + METADATA_FILENAME,
-    JSON.stringify(metadata, undefined, 2)
-  )
```
Comment (hongaar) [inline deploy-image/src/index.ts:127]: Can you check with @CalliEve what the implications are with removing this logic?

**Review by hongaar — COMMENTED** [review]
Suggest to make this action registry agnostic. Could be implemented by adding a `registry` input which defaults to `"registry-1.docker.io"` and always prefixing tags.

The name of this action suggests it doesn't build the image, maybe rename to `build-push-image`? Or `manage-image` if we also add the delete logic back to this action?

File: package.json (hunk 1 above)
Comment (ErnstWevers) [reply package.json:13]: I called it `push-image-ghcr` and specifically set it up to push images with a tag of <BRANCH_NAME> to GHCR. 

My proposal is to make a second action with a name like `push-image-dockerhub` that handles pushing realeases and their corresponding tags to dockerhub.

A build-image action sounds like a good addition, I'm not sure about persisting it without pushing though. I do like the idea of moving to smaller actions with less inputs. And defining more complex sequences of tasks in the workflows.

Comment (ErnstWevers) [reply package.json:13]: fixed the ordinality

File: README.md (hunk 2 above)
Comment (ErnstWevers) [reply README.md:11]: fixed the ordinality

File: README.md (hunk 3 above)
Comment (ErnstWevers) [reply README.md:126]: done

File: package.json (hunk 1 above)
Comment (hongaar) [reply package.json:13]: I'm not a fan of two actions doing the same thing only with different parameters. Registry feels more like a variable than a different function IMO. This also allows us to change e.g. the default registry without having to reach into each individual component workflow.

File: push-image-ghcr/action.yml (hunk 4 above)
Comment (ErnstWevers) [reply push-image-ghcr/action.yml:3]: done

File: push-image-ghcr/src/index.ts (hunk 5 above)
Comment (ErnstWevers) [reply push-image-ghcr/src/index.ts:17]: I not sure what to align here

File: push-image-ghcr/src/metadata.ts (hunk 7 above)
Comment (ErnstWevers) [reply push-image-ghcr/src/metadata.ts:5]: done

File: push-image-ghcr/src/metadata.ts (hunk 9 above)
Comment (ErnstWevers) [reply push-image-ghcr/src/metadata.ts:30]: done

File: deploy-image/src/index.ts (hunk 10 above)
Comment (ErnstWevers) [reply deploy-image/src/index.ts:99]: we discussed this, I think we can remember it's somewhere in the git tree and more on.

File: deploy-image/src/index.ts (hunk 11 above)
Comment (ErnstWevers) [reply deploy-image/src/index.ts:127]: I have re-added the logic, I does have a dependency to the `Dockerfile` in the component repo though.

File: package.json (hunk 1 above)
Comment (hongaar) [reply package.json:13]: > My proposal is to make a second action with a name like push-image-dockerhub that handles pushing realeases and their corresponding tags to dockerhub.

I'm not a fan of two actions doing the same thing only with different parameters. Registry feels more like a variable than a different function IMO. This also allows us to change e.g. the default registry without having to reach into each individual component workflow.

> A build-image action sounds like a good addition, I'm not sure about persisting it without pushing though. I do like the idea of moving to smaller actions with less inputs. And defining more complex sequences of tasks in the workflows.

A good level of abstraction for me is the level of detail consumers would normally be concerned with (e.g. the `db` action which does several steps but consumers are only interested in the compound effect, not the steps). Is there a use case where components want to separate building and pushing images? Alternatively we could have separate actions and one 'composite' action tying these together for easier consumption (e.g. https://github.com/exivity/actions/blob/main/process-binary/action.yml)

File: push-image-ghcr/src/index.ts (hunk 5 above)
Comment (hongaar) [reply push-image-ghcr/src/index.ts:17]: My bad, should have commented on the next line

```
const component = getInput('component') || defaultComponent
```

The `"component"` input is not documented

File: deploy-image/src/index.ts (hunk 11 above)
Comment (hongaar) [reply deploy-image/src/index.ts:127]: 👍 Please add something in the README to document this

**Comment by hongaar** [comment]
@ErnstWevers I build upon your work and merged with #230. The most notable change is the abstraction of the Docker registry. Internally the login, build and push (and also the metadata.json generator) has been separated.

**Pull request closed without merging**